pkg/generate/app/cmd/newapp.go

```diff
@@ -629,99 +589,16 @@ func (c *AppConfig) buildPipelines(components app.ComponentReferences, environme
-				glog.V(4).Infof("will use %q as the base image for a source build of %q", ref, refInput.Uses)
-				input, err := app.InputImageFromMatch(refInput.ResolvedMatch)
-				if err != nil {
-					return nil, fmt.Errorf("can't build %q: %v", refInput, err)
-				}
-				if !input.AsImageStream {
-					glog.Warningf("Could not find an image stream match for %q. Make sure that a Docker image with that tag is available on the node for the build to succeed.", refInput.ResolvedMatch.Value)
-				}
-
-				strategy, source, err := app.StrategyAndSourceForRepository(refInput.Uses, input)
-				if err != nil {
-					return nil, fmt.Errorf("can't build %q: %v", refInput, err)
-				}
-
-				// Override resource names from the cli
-				name = c.Name
-				if len(name) == 0 {
-					var ok bool
-					name, ok = (app.NameSuggestions{source, input}).SuggestName()
-					if !ok {
-						return nil, fmt.Errorf("can't suggest a valid name, please specify a name with --name")
-					}
-				}
-				name, err = ensureValidUniqueName(names, name)
-				source.Name = name
-				if err != nil {
-					return nil, err
-				}
-
-				// Append any exposed ports from Dockerfile to input image
-				if refInput.Uses.IsDockerBuild() && refInput.Uses.Info() != nil {
-					node := refInput.Uses.Info().Dockerfile.AST()
-					ports := dockerfileutil.LastExposedPorts(node)
-					if len(ports) > 0 {
-						if input.Info == nil {
-							input.Info = &imageapi.DockerImage{
-								Config: &imageapi.DockerConfig{},
-							}
-						}
-						input.Info.Config.ExposedPorts = map[string]struct{}{}
-						for _, p := range ports {
-							input.Info.Config.ExposedPorts[p] = struct{}{}
-						}
-					}
-				}
-				if pipeline, err = app.NewBuildPipeline(refInput.String(), input, c.OutputDocker, strategy, c.GetBuildEnvironment(environment), source); err != nil {
-					return nil, fmt.Errorf("can't build %q: %v", refInput, err)
-				}
-
-			} else {
-				glog.V(4).Infof("will include %q", ref)
-				input, err := app.InputImageFromMatch(refInput.ResolvedMatch)
-				if err != nil {
-					return nil, fmt.Errorf("can't include %q: %v", refInput, err)
-				}
-				name = c.Name
-				if len(name) == 0 {
-					var ok bool
-					name, ok = input.SuggestName()
-					if !ok {
-						return nil, fmt.Errorf("can't suggest a valid name, please specify a name with --name")
-					}
-				}
-				name, err = ensureValidUniqueName(names, name)
-				if err != nil {
-					return nil, err
-				}
-				input.ObjectName = name
-				if pipeline, err = app.NewImagePipeline(refInput.String(), input); err != nil {
-					return nil, fmt.Errorf("can't include %q: %v", refInput, err)
-				}
-			}
-
-			if c.Deploy {
-				if err := pipeline.NeedsDeployment(environment, c.Labels, name); err != nil {
-					return nil, fmt.Errorf("can't set up a deployment for %q: %v", refInput, err)
-				}
+			pipeline, err := app.NewPipeline(c.Name, names, ref, environment, c.AddEnvironmentToBuild, c.Labels, c.OutputDocker, c.Deploy)
```

Comment:
Just a suggestion... the code might look cleaner if you have an object called PipelineConfig (or whatever you want to call it) that has fields for the name generator, environment, and all the other config you are passing here... create it once at the beginning of the function, and invoke a NewPipeline(ref) method on it. Since ref is the only thing that's going to change as you iterate through the pipelines. 

Similarly for name generation, I would create a NameGenerator object (field on PipelineConfig) that internally keeps track of the names map and provides a method like ResourceName(name, suggester) to give you back a name.